Export C++ standard selection in pkg-config files

doudou opened the pull request.

2maz commented on modules/Rock.cmake:
```diff
@@ -500,6 +504,18 @@ function(rock_prepare_pkgconfig TARGET_NAME DO_INSTALL)
     set(PKGCONFIG_CFLAGS ${${TARGET_NAME}_PKGCONFIG_CFLAGS})
     set(PKGCONFIG_LIBS ${${TARGET_NAME}_PKGCONFIG_LIBS})
 
+    if (DEFINED ROCK_PUBLIC_CXX_STANDARD)
```
Is there documentation for this variable already?

doudou replied:
I'll properly document `rock_activate_cxx11`

2maz commented on modules/Rock.cmake:
```diff
@@ -500,6 +504,18 @@ function(rock_prepare_pkgconfig TARGET_NAME DO_INSTALL)
 
+    if (DEFINED ROCK_PUBLIC_CXX_STANDARD)
+        set(cxx_standard ${ROCK_PUBLIC_CXX_STANDARD})
+    elseif (CMAKE_VERSION VERSION_LESS "3.1")
```
I do not see the real benefit of two ways for setting the cxx_standard. How likely to build targets with different cxx_standards in the same project in our environments?

doudou replied:
The `ROCK_PUBLIC_CXX_STANDARD` variable sets the standard as exported in the .pc file, it does not change how the standard is handled internally to the package. It is meant as a way to use C++11 internally but have C++98 headers (and avoid propagating the C++11 choice downstream).

Note that for CMake >= 3.1, when ROCK_PUBLIC_CXX_STANDARD is not used, the code only relies on CMake's own way to handle the C++ standard (setting the CXX_STANDARD property on the targets).

2maz replied:
I suggest you add this exact explanation to the documentation as well

doudou replied:
I rephrased it, but I thought I added the same information to the rock_activate_cxx11 macro documentation. No ?

2maz replied:
Your description (first part) here is a bit more explicit on the target use case.

doudou replied:
Updated. Is that better ?

2maz replied:
Thx, yes that is better.

doudou commented:
Updated with documentation of the C++ selection mechanism on rock_activate_cxx11.

I've added a section about using the CMake native mechanisms at the end, and even recommend it. Comments ?

2maz commented:
See comment on the variable doc, otherwise feel good to merge.

jmachowinski commented:
This breaks existing code, if no pc.in file exists for a library.
Either fix or revert.

planthaber commented:
Also, no header only libraries are possible anymore:

> CMake Error at /home/dfki.uni-bremen.de/planthaber/dfki/rock/entern/install/share/rock/cmake/Rock.cmake:542 (get_property):
>   get_property could not find TARGET osgVizInterfaces.  Perhaps it has not
>   yet been created.

Which is true as no target was created

2maz commented:
@jmachowinski: I cannot reproduce an error related to a missing pc.in -- any setup example or is this the same as @planthaber reported?

@planthaber: ack, this might be a common pattern for (ab)using rock_library (which defines a shared library) thus misses a warning at least and now probably a deprecation mechanism as well: please use 'rock_install_headers' for header only installation.

planthaber commented:
The lib is the same, my error occured when the .pc.in file was added.

rock_install_headers results in this error:
> - Install configuration: "RelWithDebInfo"
> CMake Error at src/cmake_install.cmake:36 (file):
>   file INSTALL cannot find
>   "/home/dfki.uni-bremen.de/planthaber/dfki/rock/entern/gui/osgviz/osgviz/src/osgVizInterfaces".
> Call Stack (most recent call first):
>   cmake_install.cmake:37 (include)

2maz commented:
Apparently I missed that pc.in must be installed as well, so rock_install_headers is not a sufficient replacement, but requires a fix of rock_prepare_pkgconfig: @planthaber: pls have a look at #44

2maz commented on modules/Rock.cmake:
```diff
@@ -500,6 +534,18 @@ function(rock_prepare_pkgconfig TARGET_NAME DO_INSTALL)
+    endif()
+
+    if (cxx_standard)
+        set(PKGCONFIG_CFLAGS "${PKGCONFIG_CFLAGS} -std=c++${cxx_standard}")
```
Since CXX_EXTENSIONS will be on by default the hard setting of -std=c++11 and forwarding of this flag via pkg-config can conflict with the actual used compile flag -gnu++11, e.g. triggering issues like base-numeric#18 

(see also https://cmake.org/cmake/help/v3.1/manual/cmake-compile-features.7.html#manual:cmake-compile-features(7) for reference)

Obviously one option would be to disable CXX_EXTENSIONS by default. @doudou: comments?

2maz replied:
added #45 for further discussion

doudou commented:
FYI. I used the opportunity of changing base/cmake to create a testsuite for inter-package build problems (which is mainly aimed at the cmake macros, but is also relevant for interactions between libraries and orogen)

Just build [this buildconf](https://github.com/rock-core/rock.build-tests-buildconf.git). All related packages are prefixed with `build_tests-` on github (and are in build_tests/ on disk)

Beyond the build itself, more involved tests are present in `build_tests/test`.

I've just added two test packages for the headers-only issue. The build currently does not pass because of these bugs (good !). Unfortunately, orogen does pass because by default gcc will add C++11 and C++14 extensions, so it's really hard to get an error if the flags are wrong. I currently have a patched typelib just to get the orogen packages to fail.

doudou commented:
> Unfortunately, orogen does pass because by default gcc will add C++11 and C++14 extensions, so it's really hard to get an error if the flags are wrong. I currently have a patched typelib just to get the orogen packages to fail.

Actually, fixed that by testing for `-std=c++11` using the preprocessor instead.

2maz commented:
Just tried the buildconf: 

utilrb: branch pkgconfig_export_requires is unknown

and with a shot into the blue by removing the utilrb override:
```
build_tests/cmake/rock_library_no_public_dependencies(/opt/workspace/test/build_tests/cmake/rock_library_no_public_dependencies): failed in configure phase

Unknown CMake command "rock_no_public_dependencies".

doudou commented:
The `export_cxx11_in_pkgconfig_files` branch now contains a hopefully fixed version of the changes. I unfortunately can't create a new pull request, as github does not understand a `merge --ours` commit - i.e. believes `master` contains all the commits in the branch.

@planthaber since you were experiencing issues with the original patchset, could you please test ?

doudou commented:
> The export_cxx11_in_pkgconfig_files branch now contains a hopefully fixed version of the changes. I unfortunately can't create a new pull request, as github does not understand a merge --ours commit - i.e. believes master contains all the commits in the branch.
> 
> @planthaber since you were experiencing issues with the original patchset, could you please test ?

Ping ...

doudou commented:
OK .. .given the amount of attention this is getting, I'm merging the hopefully fixed branch. At least if something breaks, it should make you react.
